File: denseSnp.py
import pandas as pd
import sys

# user input
myfile = sys.argv[1]
#myfile1="CEPH_chips5-8_FinalReport.txt"
myfile1="CEPH_chips_9-12_FinalReport.txt"
# import data
#mydata = pd.read_csv('/home/data/LabData/SnpArrayData/Coriell/CEPH_chips_1_4/CEPH_chips1_4_full_data.table.txt.gz',sep='\t', skiprows=9, usecols=[1,18,19,0,10,11], dtype={'Sample ID':str,'Chr':str,'Position':int,'SNP Name':str,'Allele1 - Forward':str,'Allele2 - Forward':str} )
#mydata = pd.read_csv(myfile,sep='\t', skiprows=9, usecols=[1,18,19,0,10,11,4,20,*range(28,33),35,36], dtype={'SNP Name':str,\
mydata = pd.read_csv(myfile,sep='\t', skiprows=9, dtype={\
'SNP Name':str,\
'Sample ID':str,\
'Allele1 - Top':str,\
'Allele2 - Top':str,\
'GC Score':float,\
'Sample Name':str,\
'Sample Group':str,\
'Sample Index':str,\
'SNP Index':float,\
'SNP Aux':float,\
'Allele1 - Forward':str,\
'Allele2 - Forward':str,\
'Allele1 - Design':str,\
'Allele2 - Design':str,\
'Allele1 - AB':str,\
'Allele2 - AB':str,\
'Allele1 - Plus':str,\
'Allele2 - Plus':str,\
'Chr':str,\
'Position':int,\
'GT Score':float,\
'Cluster Sep':float,\
'SNP':str,\
'ILMN Strand':str,\
'Customer Strand':str,\
'Top Genomic Sequence':str,\
'Plus/Minus Strand':str,\
'Theta':float,\
'R':float,\
'X':float,\
'Y':float,\
'X Raw':float,\
'Y Raw':float,\
'B Allele Freq':float,\
'Log R Ratio':float,\
'CNV Value':float,\
'CNV Confidence':float} )

# checking type
type(mydata)

# for samtools faidx
mydata.loc[:,'region_samtools']= ("chr"+mydata['Chr']+":"+\
mydata['Position'].map(str)+"-"+\
mydata['Position'].map(str))

# filtering uninformative fields
mydata_filtered=mydata[(mydata.Chr !='XY') & (mydata.Chr!="0") & (mydata['Allele1 - Forward'] !="-") & (mydata.Chr !='MT')]

# Uncomment the following. Was commented out to avoid regenerating the same file again.
mydata_chrRegions= mydata_filtered['region_samtools']

# file for samtools faidx
mydata_chrRegions.to_csv(myfile1+'_regions_samtools.txt',index=False,header=False)

# output tidy file
mydata_filtered.to_csv(myfile1+'_tidy_data.txt',index=False,header=True)

# to August
#mydata_to_August = mydata_filtered[['Sample ID','Chr','Position','Allele1 - Forward','Allele2 - Forward','GC Score','GT Score','R','X','Y','X Raw','Y Raw','CNV Value','CNV Confidence']]
#mydata_to_August.to_csv(myfile1+'_tidy_data_to_August.tsv', index=False,header=True,sep="\t")

File: denseSnp.R
library(tidyverse)
library(data.table)

# complement funciton
complement <- function(x){
  case_when(
    x=="A" ~ "T",
    x=="T" ~ "A",
    x=="G" ~ "C",
    x=="C" ~ "G"
  )
}


#mytidydata_input <- read_csv("CEPH_chips1_4_full_data.table.txt.gz_tidy_data.txt", col_names = TRUE)
#mytidydata_input <- fread("CEPH_chips1_4_full_data.table.txt.gz_tidy_data.txt")
#mytidydata_input <- fread("CEPH_chips5-8_FinalReport.txt_tidy_data.txt")
mytidydata_input <- fread("CEPH_chips_9-12_FinalReport.txt_tidy_data.txt")

#system.time({mytidydata_input <- fread("CEPH_chips1_4_full_data.table.txt.gz_tidy_data.txt")})

#myreffasta<-read_delim("regions_variants_ref_oneline.fasta", col_names = c('region_samtools','ref_base'), delim = "\t")
myreffasta<-read_delim("regions_variants_chips_9_12_ref.fastaoneline.fasta", col_names = c('region_samtools','ref_base'), delim = "\t")
setDT(myreffasta)

# a problem with fread, it skips the first row
#myreffasta<-fread("regions_variants_ref_oneline.fasta", col.names = c('region_samtools','ref_base'))
#myreffasta<-fread("regions_variants_chips_5-8_ref.fastaoneline.fasta", col.names = c('region_samtools','ref_base'))

#setkey(myreffasta)

myreffasta_1 <- myreffasta %>% 
  separate(region_samtools, into = c('foo','region_samtools'), remove = TRUE, sep = ">") %>% 
  select(-foo)

myreffasta_2 <- myreffasta_1 %>% distinct()

# following is an intermediate step as the size of the objects are too big
mytidydata_input_join <- mytidydata_input %>% 
  left_join(myreffasta_2, by=c("region_samtools"="region_samtools")) %>% 
  distinct()


# skip if you don't want intermediate outputs
#write_csv(mytidydata_input_join,path = "mytidydata_join.csv")
#mytidydata_foo <- read_delim("mytidydata_join.csv", delim=",")
#mytidydata <- fread("mytidydata_join.csv")

# following is an intermediate step as the size of the objects are too big
mytidydata <- mytidydata_input_join

# some clean up to emtpy space
rm(mytidydata_input_join,myreffasta,mytidydata_input,myreffasta_1,myreffasta_2)

################################################################################
## Testing
# mytidydata %>% head()
# 
# #setDT(mytidydata)[,]
# 
# mytidydata %>% 
#   group_by(`Sample ID`, Position, Chr) %>% 
#   summarise(dup = n()) %>% 
#   arrange(desc(dup))
# 
# mytidydata_500K %>% 
#   group_by(`Sample ID`, Position, Chr) %>% 
#   mutate(dup = n()) -> mytidydata_1
# 
# mytidydata_1 %>% 
#   filter(dup > 1) -> mytidydata_2
# 
# # to see if there are any dup snps
# mytidydata_2 %>% ungroup() %>% 
#   group_by(`Sample ID`,Position, Chr) %>% 
#   arrange(Chr, Position, desc(dup)) -> mytidydata_3
# 
# # maybe this info is useful?
# infiStrand <- read_delim("InfiniumOmni5-4v1-2_A1-b37.Ilmn.strand", delim = "\t",  col_names = F)

#################################################################################

#setDT(mytidydata)

# mytidydata %>% head(n=1000000) -> mytidydata_500K # test data set

mytidydata %>% 
  select(-region_samtools) %>%
  separate(SNP, into = c("foo1","foo2"),sep = "[/]", remove = F) %>% 
  separate(foo1, into = c(NA, "Snp1"),sep = -1) %>% 
  separate(foo2, into = c("Snp2", NA), sep=-1) -> mytidydata

# get the forward alleles comparing it to the reference allele
#system.time(
mytidydata %>% 
  #select(`SNP Name`:Position,SNP:`Plus/Minus Strand`,ref_base) %>% #filtered for test purpose
  mutate(FrwdOrRevCalculation=ifelse(ref_base == `Allele1 - Forward` | ref_base == `Allele2 - Forward`, 
                                    "Frwd",
                                    ifelse(ref_base == complement(`Allele1 - Forward`) | ref_base == complement(`Allele2 - Forward`), 
                                           "Complement",
                                            ifelse(ref_base == Snp1, 
                                                   case_when(Snp2==`Allele1 - Forward` | Snp2==`Allele2 - Forward` ~ "Frwd",
                                                             complement(Snp2)==`Allele1 - Forward` | complement(Snp2)==`Allele2 - Forward` ~ "Complement"),
                                                   ifelse(ref_base == complement(Snp1),
                                                          case_when(complement(Snp2)==`Allele1 - Forward` | complement(Snp2)==`Allele2 - Forward` ~ "Frwd",
                                                                    Snp2==`Allele1 - Forward` | Snp2==`Allele2 - Forward` ~ "Complement"),
                                                          ifelse(ref_base == Snp2,
                                                                 case_when(Snp1==`Allele1 - Forward` | Snp1==`Allele2 - Forward` ~ "Frwd",
                                                                           complement(Snp1) ==`Allele1 - Forward` | complement(Snp1)==`Allele2 - Forward` ~ "Complement"),
                                                                 ifelse(ref_base==complement(Snp2),
                                                                        case_when(complement(Snp1)==`Allele1 - Forward` | complement(Snp1)==`Allele2 - Forward` ~ "Frwd",
                                                                                  Snp1==`Allele1 - Forward` | Snp1==`Allele2 - Forward` ~ "Complement"
                                                                                  ),
                                                                        "SomethingisWrong"
                                                                        )
                                                                 )
                                                          )
                                                   )       
                                    )
                        )
                        ) %>% 
  mutate(TrueAllele1_Forward = ifelse(FrwdOrRevCalculation=="Frwd",`Allele1 - Forward`,complement(`Allele1 - Forward`)),
         TrueAllele2_Forward = ifelse(FrwdOrRevCalculation=="Frwd",`Allele2 - Forward`,complement(`Allele2 - Forward`))
         ) %>%
  select(`SNP Name`,`Sample ID`,TrueAllele1_Forward,TrueAllele2_Forward,
         ref_base,`Allele1 - AB`,`Allele2 - AB`,`GC Score`:`SNP Aux`,
         Chr:`CNV Confidence`,everything())-> mytidydata_results#)

setDT(mytidydata_results)
#fwrite(mytidydata_results, file = "mytidydata_final_results.txt", sep = "\t")
fwrite(mytidydata_results, file = "CEPH_chips_9-12_FinalReport.final_results.txt", sep = "\t", buffMB = 1024L)
